### PR TITLE
Generalized expectation derivative

### LINKED ISSUE #148: Derivative of QuantumInference expectation value method for general functions
Using the [`get_gradient_circuits`](https://www.tensorflow.org/quantum/api_docs/python/tfq/differentiators/Differentiator#get_gradient_circuits) function in the TFQ differentiators, we can take the derivative of expectation values of general functions.

### PR DESCRIPTION
**Pull request by zaqqwerty**
Implement a custom gradient for `_sampled_expectation` to enable derivatives of `Hamiltonian`s with general `BitstringEnergy` diagonals.

Resolves #148

### CODE REVIEW

File: qhbmlib/inference/qnn.py [hunk 1]
```diff
@@ -121,22 +121,95 @@ def _sampled_expectation(self, initial_states: tf.Tensor,
...
-        samples,
-        fn_output_signature=tf.float32)
-    return utils.expand_unique_results(unique_expectations, idx)
+
+    @tf.custom_gradient
+    def _inner_expectation():
+      """Enables derivatives."""
+      unique_states, idx, _ = utils.unique_bitstrings_with_counts(
+          initial_states)
+      u = self.circuit + observable.circuit_dagger
```
Comment (farice) [inline qhbmlib/inference/qnn.py:130]: nit: this is not necessarily unitary so wouldn't call it `u`

File: tests/inference/qnn_test.py [hunk 2]
```diff
@@ -425,12 +428,20 @@ def test_expectation_bitstring_energy(self):
...
         p=act_fraction)
     state_raw_circuit = state_raw_circuits[0]
-    state_circuit = models.DirectQuantumCircuit(state_raw_circuit)
+    qnn_minval = 0.25
+    qnn_maxval = 0.75
+    state_circuit_initializer = tf.keras.initializers.RandomUniform(
+        minval=qnn_minval, maxval=qnn_maxval, seed=self.tf_random_seed)
+    state_circuit = models.DirectQuantumCircuit(state_raw_circuit,
+                                                state_circuit_initializer)
+    state_circuit.build([])
```
Comment (farice) [inline tests/inference/qnn_test.py:437]: jfme: why do we have to call build here now?

File: tests/inference/qnn_test.py [hunk 3]
```diff
@@ -441,22 +452,32 @@ def test_expectation_bitstring_energy(self):
...
+    hamiltonian_circuit = models.DirectQuantumCircuit(
+        hamiltonian_raw_circuit, hamiltonian_circuit_initializer)
+    hamiltonian_circuit.build([])
 
     # Total circuit
     bitstring_circuit = models.circuit_utils.bit_circuit(qubits)
     measurement_circuit = cirq.Circuit(
         cirq.measure(q, key=f"measure_qubit_{n}") for n, q in enumerate(qubits))
+    # needed later for retrieving sample results from Cirq output
+    qb_keys = [(q, f"measure_qubit_{i}") for i, q in enumerate(qubits)]
```
Comment (farice) [inline tests/inference/qnn_test.py:466]: `qb_keys` doesn't appear to be used?

File: tests/inference/qnn_test.py [hunk 4]
```diff
@@ -475,41 +496,125 @@ def generate_resolvers():
...
+    num_layers = 1
+    random.seed(self.python_random_seed)
     bits = random.sample(range(1000), self.num_bits)
-    units = random.sample(range(1, 100), num_layers)
-    activations = random.sample([
-        "elu", "exponential", "gelu", "hard_sigmoid", "linear", "relu", "selu",
-        "sigmoid", "softmax", "softplus", "softsign", "swish", "tanh"
-    ], num_layers)
+    units = [2] * num_layers
+    activations = ["linear"] * num_layers
```
Comment (farice) [inline tests/inference/qnn_test.py:503]: why only linear now?

File: tests/inference/qnn_test.py [hunk 5]
```diff
@@ -475,41 +496,125 @@ def generate_resolvers():
...
         actual_expectations, expected_expectations, rtol=self.close_rtol)
+    self.assertAllEqual(
+        tf.shape(actual_expectations), [len(initial_states_list), 1])
+
+    # TODO(#205) function seems to be ripe for refactoring
+    def delta_expectations_func(k, var, delta):
+      """Calculate the expectation with kth entry of `var` perturbed."""
+      num_elts = tf.size(var)
+      old_value = var.read_value()
+      # Can no longer assume var is 1D, since energy uses matrix variables now
```
Comment (farice) [inline tests/inference/qnn_test.py:568]: can we use this adapted version of the fn for `test_expectation_cirq`?

File: qhbmlib/inference/qnn.py [hunk 6]
```diff
@@ -121,22 +121,95 @@ def _sampled_expectation(self, initial_states: tf.Tensor,
...
+            variables,
+            output_gradients=upstream[0],
+            unconnected_gradients=tf.UnconnectedGradients.ZERO)
+
+        # This block adapted from my `differentiate_sampled` in TFQ.
+        (batch_programs, new_symbol_names, batch_symbol_values, batch_weights,
+         batch_mapper) = self.differentiator.get_gradient_circuits(
+             unique_circuits, u.symbol_names, unique_tiled_values)
+        m_i = tf.shape(batch_programs)[1]
+        n_ops = 1
```
Comment (farice) [inline qhbmlib/inference/qnn.py:161]: var is only used once so doesn't require a name

File: qhbmlib/inference/qnn.py [hunk 7]
```diff
@@ -121,22 +121,95 @@ def _sampled_expectation(self, initial_states: tf.Tensor,
...
+        unique_expectations = tf.map_fn(
+            lambda x: tf.math.reduce_mean(observable.energy(x)),
+            unique_samples,
+            fn_output_signature=tf.float32)
+        forward_pass = tf.expand_dims(
+            utils.expand_unique_results(unique_expectations, idx), 1)
+
+      def grad_fn(*upstream, variables):
+        """Use `get_gradient_circuits` method to get QNN variable derivatives"""
+        thetas_gradients = thetas_tape.gradient(
```
Comment (farice) [inline qhbmlib/inference/qnn.py:150]: nit: would move this to the bottom of grad_fn since it sits unused for so many lines

File: qhbmlib/inference/qnn.py (hunk 1 above)
Comment (zaqqwerty) [reply qhbmlib/inference/qnn.py:130]: We do assume it is unitary, otherwise the gradient result is not correct.  But I agree `u` is not an informative name, changed to `total_circuit`

File: qhbmlib/inference/qnn.py (hunk 7 above)
Comment (zaqqwerty) [reply qhbmlib/inference/qnn.py:150]: Done

File: tests/inference/qnn_test.py (hunk 3 above)
Comment (zaqqwerty) [reply tests/inference/qnn_test.py:466]: Used below but accidentally created the same `qb_keys` twice, now removed the second copy.  I decided to put `qb_keys` creation here rather than where it is used because `qb_keys` needs to use the same measurement keys as the `measurement_circuit` above it.  Added a comment describing this.

File: tests/inference/qnn_test.py (hunk 5 above)
Comment (zaqqwerty) [reply tests/inference/qnn_test.py:568]: The code there is slightly different right now, but I think we can refactor this out eventually.   Hence #205

File: tests/inference/qnn_test.py (hunk 2 above)
Comment (zaqqwerty) [reply tests/inference/qnn_test.py:437]: Removed, tested it and turns out we don't need this one

File: qhbmlib/inference/qnn.py (hunk 1 above)
Comment (farice) [reply qhbmlib/inference/qnn.py:130]: The observable can't be a general Hermitian operator?

File: tests/inference/qnn_test.py (hunk 5 above)
Comment (farice) [reply tests/inference/qnn_test.py:568]: I would bias towards not introducing more clutter and taking out another TODO loan. You're call

File: tests/inference/qnn_test.py (hunk 4 above)
Comment (zaqqwerty) [reply tests/inference/qnn_test.py:503]: Restricted during debugging and forgot to add things back.  Restored to all activations now.

File: qhbmlib/inference/qnn.py (hunk 1 above)
Comment (zaqqwerty) [reply qhbmlib/inference/qnn.py]: We assume the operator is diagonalized into a unitary circuit (`observable.circuit`) and a diagonal (`observable.energy`).  Every Hermitian operator can be represented this way, so it is fully general

File: tests/inference/qnn_test.py (hunk 3 above)
Comment (farice) [reply tests/inference/qnn_test.py:466]: the usual heuristic is to move variable construction to the "latest possible line" so that it's used shortly after it's created. In this case, it appears you can move both `qb_keys` and `measurement_circuit` much later

File: tests/inference/qnn_test.py (hunk 5 above)
Comment (zaqqwerty) [reply tests/inference/qnn_test.py:570]: The code is different thought, so n my opinion it would be more cluttered to do the refactoring now.  Properly it should be a test util function that is tested, then we can confidently use a generic version (similar code also shows up in `ebm_test.py`, agains slightly different).  In my opinion the proper generic version is a function that accepts the expectation value producing function as an argument, that we would `functools.partial` away in each test

**Review by farice — CHANGES_REQUESTED** [review]
Thanks for the quick reply - please confirm whether the observable is strongly restricted to being a unitary operator as opposed to a general Hermitian operator

File: qhbmlib/inference/qnn.py (hunk 1 above)
Comment (farice) [reply qhbmlib/inference/qnn.py]: Thanks for clarifying - it means that the observable is represented as QHBM. I missed this aspect again where it seemed like a class could be any general representation and not necessarily spectral

Comment (farice) [reply qhbmlib/inference/qnn.py]: My suggestion would be to at least update the docstring with the parameterization specified

**Comment by zaqqwerty** [comment]
It accepts every Hermitian operator, the restriction is only on the representation of the operator

File: tests/inference/qnn_test.py (hunk 5 above)
Comment (farice) [reply tests/inference/qnn_test.py:570]: Makes sense - will be a high yield TODO for readability then IMO

File: tests/inference/qnn_test.py [hunk 8]
```diff
@@ -40,12 +40,14 @@ class QuantumInferenceTest(parameterized.TestCase, tf.test.TestCase):
   def setUp(self):
     """Initializes test objects."""
     super().setUp()
+    self.python_random_seed = 11
     self.tf_random_seed = 10
+    self.tf_random_seed_alt = 201
```
Comment (farice) [inline tests/inference/qnn_test.py:45]: does this mean we need two tf seeds for the tests to all pass?

File: tests/inference/qnn_test.py [hunk 9]
```diff
@@ -441,22 +451,33 @@ def test_expectation_bitstring_energy(self):
         n_moments=n_moments,
         p=act_fraction)
     hamiltonian_raw_circuit = hamiltonian_raw_circuits[0]
-    hamiltonian_circuit = models.DirectQuantumCircuit(hamiltonian_raw_circuit)
+    hamiltonian_circuit_initializer = tf.keras.initializers.RandomUniform(
+        minval=qnn_minval, maxval=qnn_maxval, seed=self.tf_random_seed_alt)
+    hamiltonian_circuit = models.DirectQuantumCircuit(
+        hamiltonian_raw_circuit, hamiltonian_circuit_initializer)
+    hamiltonian_circuit.build([])
```
Comment (farice) [inline tests/inference/qnn_test.py:458]: is this build needed? is the reason we need it now so that the initializer is constructed? is this something we can eventually automate, like before?

File: tests/inference/qnn_test.py (hunk 3 above)
Comment (zaqqwerty) [reply tests/inference/qnn_test.py:466]: If we move `measurement_circuit` later, then we also need to move `total_circuit` later.  But then `state_raw_circuit` and `hamiltonian_raw_circuit` are used farther away from where they were created.  The grouping of having `total_circuit` defined here was to keep all quantum circuit set up more local

File: tests/inference/qnn_test.py [hunk 10]
```diff
@@ -475,41 +496,127 @@ def generate_resolvers():
       } for r in bitstring_resolvers]
 
     # hamiltonian energy
-    num_layers = 5
+    num_layers = 1
```
Comment (farice) [inline tests/inference/qnn_test.py:499]: jfme: why decrease to a single layer? to make the test faster?

File: tests/inference/qnn_test.py (hunk 3 above)
Comment (farice) [reply tests/inference/qnn_test.py:466]: I see - I would still move `qb_keys` to the bottom since it's absent of these dependencies and strictly makes usages more local. Is it clear why this heuristic is more readable?

File: tests/inference/qnn_test.py (hunk 9 above)
Comment (farice) [reply tests/inference/qnn_test.py:458]: definitely fair to build and keep as-is - i recall you had a PR which automated building of member classes. that was different because you built members rather than the class itself which makes for a reasonable distinction

File: tests/inference/qnn_test.py (hunk 10 above)
Comment (zaqqwerty) [reply tests/inference/qnn_test.py:499]: Yes, was to make the tests go faster.  Same with number of units, since the finite difference derivatives are what eat most of the time now.  Not too bad to bump it to 2 so I did that.

File: tests/inference/qnn_test.py (hunk 3 above)
Comment (zaqqwerty) [reply tests/inference/qnn_test.py]: Yes, the localization heuristic makes things more readable in general.  Depending on the situation though, I think deviations from this heuristic can be more readable - like here.  But moving just `qb_keys` later, rather than both `qb_keys` and `measurement_circuit`, is simple enough - done.

File: tests/inference/qnn_test.py (hunk 8 above)
Comment (zaqqwerty) [reply tests/inference/qnn_test.py:45]: Added as a comment

File: tests/inference/qnn_test.py (hunk 3 above)
Comment (farice) [reply tests/inference/qnn_test.py]: Thank you! In the future, it's completely reasonable to stick to your personal preference. This was just mine, independent of the heuristic.
